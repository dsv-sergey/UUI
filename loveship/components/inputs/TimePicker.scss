@use '../../assets/styles/scss/index'  as *;

.root {
    &:global(.uui-timepicker-container) {
        display: flex;
        padding: 8px;

        :global(.uui-timepicker-input) {
            &:global(.uui-input-box) {
                border: 1px solid $night600;
                width: 36px;
                height: 36px;
            }

            > input {
                text-align: center;
                color: $night700;
                font-family: $font-sans;
            }
        }

        :global(.uui-timepicker-item) {
            position: relative;
            padding: 5px;
        }

        :global(.uui-timepicker-item:first-child) {
            &::after {
                color: $night600;
                font-size: 20px;
                content: ':';
                position: absolute;
                top: 33px;
                left: 45px;
            }
        }

        .icon {
            align-items: center;
            fill: $night600;

            &:hover {
                cursor: pointer;
            }
        }

        :global(.uui-timepicker-icon-up) {
            align-items: center;
            fill: $night600;
            transform: rotate(180deg);
        }

        :global(.uui-timepicker-icon-down) {
            align-items: center;
            fill: $night600;
        }
    }
}

.date-input {
    width: 100%;
}

:global(.uui-dropdown-body) {
    align-items: center;
}
