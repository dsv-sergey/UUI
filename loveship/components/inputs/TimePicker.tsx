import React from 'react';
import { EditMode, SizeMod } from '../types';
import moment from 'moment';
import { IEditable, IDisableable, TimePickerValue, ICanBeReadonly } from '@epam/uui';
import { TextInput } from './TextInput';
import { DropdownContainer, Dropdown } from '../overlays';
import { TimePickerBody } from './TimePickerBody';
import * as css from './TimePicker.scss';

interface TimePickerState {
    isOpen: boolean;
    value: string;
}

const valueToTimeString = (value: TimePickerValue, format: TimePickerProps['format']) => {
    return moment(value).format(format === 24 ? "HH:mm" : "hh:mm A");
};

export interface TimePickerProps extends IEditable<TimePickerValue>, IDisableable, SizeMod, ICanBeReadonly, EditMode {
    minutesStep?: number;
    format?: 12 | 24;
}

export class TimePicker extends React.Component<TimePickerProps, TimePickerState> {
    state = {
        isOpen: false,
        value: valueToTimeString(this.props.value, this.props.format),
    };

    componentDidUpdate(prevProps: TimePickerProps) {
        if (this.props.value !== prevProps.value) {
            this.setState({ ...this.state, value: valueToTimeString(this.props.value, this.props.format) });
        }
    }

    getFormat = () => {
        return this.props.format === 24 ? 'HH:mm' : 'hh:mm A';
    }

    onClear = () => {
        this.props.onValueChange({ hours: 0, minutes: 0 });
    }

    handleInputChange = (newValue: string) => {
        const value = moment(newValue, this.getFormat(), true);
        if (moment(newValue, this.getFormat(), true).isValid()) {
            this.props.onValueChange({ hours: value.hours(), minutes: value.minutes() });
            this.setState({ ...this.state, value: newValue });
        } else {
            this.setState({ ...this.state, value: newValue });
        }
    }

    handleBlur = () => {
        if (!moment(this.state.value, this.getFormat(), true).isValid()) {
            this.props.onValueChange(this.props.value);
            this.setState({ ...this.state, value: valueToTimeString(this.props.value, this.props.format) });
        }
    }

    render() {
        return (
            <Dropdown
                renderTarget={ (props) =>
                    <TextInput
                        { ...props }
                        size={ this.props.size || '36' }
                        isDisabled={ this.props.isDisabled }
                        isReadonly={ this.props.isReadonly }
                        isInvalid={ this.props.isInvalid }
                        cx={ css.dateInput }
                        value={ this.state.value }
                        onValueChange={ this.handleInputChange }
                        onCancel={ this.onClear }
                        onBlur={ this.handleBlur }
                        mode={ this.props.mode }
                        isDropdown={ false }
                    />
                }
                renderBody={ () =>
                     !this.props.isDisabled && !this.props.isReadonly && <DropdownContainer>
                        <TimePickerBody { ...this.props } value={ this.props.value }/>
                    </DropdownContainer> }
                onValueChange={ (opened) => this.setState({ ...this.state, isOpen: opened }) }
                value={ this.state.isOpen }
                modifiers={ {offset: {offset: '0,6px'}} }
            />
        );
    }
}