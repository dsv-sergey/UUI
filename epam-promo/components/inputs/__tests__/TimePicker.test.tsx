import React from 'react';
import { TimePicker } from '../TimePicker';
import renderer from 'react-test-renderer';

jest.mock('react-dom', () => ({
    findDOMNode: jest.fn(),
}));

describe('TimePicker', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(<TimePicker
                value={ null }
                onValueChange={ jest.fn }
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        let tree: any;
        tree = renderer
            .create(<TimePicker
                value={ { hours: 1, minutes: 5 } }
                onValueChange={ jest.fn }
                format={ 24 }
                minutesStep={ 5 }
                size='36'
                isDisabled
            />)
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});


