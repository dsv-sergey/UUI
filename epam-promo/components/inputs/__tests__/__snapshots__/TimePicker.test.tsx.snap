// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker should be rendered correctly 1`] = `
<div
  className="container uui-input-box -clickable root size-36 dateInput"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  tabIndex={-1}
>
  <input
    className="uui-input"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="Invalid date"
  />
  <div
    className="container uui-icon uui-enabled -clickable uui-icon-cancel -clickable"
    onClick={[Function]}
  >
    <svg
      className=""
      height={24}
      viewBox="0 0 12 24"
      width={12}
    >
      <use
        xlinkHref="#btn-cross-18.svg"
      />
    </svg>
  </div>
</div>
`;

exports[`TimePicker should be rendered correctly 2`] = `
<div
  className="container uui-input-box uui-disabled root size-36 dateInput"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  tabIndex={-1}
>
  <input
    className="uui-input"
    disabled={true}
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="01:05"
  />
  <div
    className="container uui-icon uui-disabled uui-icon-cancel -clickable hidden"
  >
    <svg
      className=""
      height={24}
      viewBox="0 0 12 24"
      width={12}
    >
      <use
        xlinkHref="#btn-cross-18.svg"
      />
    </svg>
  </div>
</div>
`;
